source/Nevermore/IQueryBuilder.cs

```diff
@@ -1,40 +1,85 @@
-        IQueryBuilder<TRecord> Where(string fieldName, SqlOperand operand, IEnumerable<object> values);
-        
+        IQueryBuilder<TRecord> WhereParameterised(string fieldName, UnarySqlOperand operand, Parameter parameter);
+        IQueryBuilder<TRecord> WhereParameterised(string fieldName, BinarySqlOperand operand, Parameter startValueParameter, Parameter endValueParameter);
+        IQueryBuilder<TRecord> WhereParameterised(string fieldName, ArraySqlOperand operand, IEnumerable<Parameter> parameterNames);
+
         IOrderedQueryBuilder<TRecord> OrderBy(string orderByClause);
         IOrderedQueryBuilder<TRecord> OrderByDescending(string orderByClause);
+        IQueryBuilder<TRecord> IgnoreDefaultOrderBy();
 
-        IQueryBuilder<TRecord> Join(IJoin join);
-        IQueryBuilder<TRecord> Parameter(string name, object value);
-        IQueryBuilder<TRecord> LikeParameter(string name, object value);
-        IQueryBuilder<TRecord> LikePipedParameter(string name, object value);
-        IQueryBuilder<TRecord> View(string viewName);
-        IQueryBuilder<TRecord> Table(string tableName);
-        IQueryBuilder<TRecord> Hint(string tableHint);
-        IQueryBuilder<TRecord> NoLock();
+        IQueryBuilder<TRecord> Column(string name);
+        IQueryBuilder<TRecord> Column(string name, string columnAlias);
+        IQueryBuilder<TRecord> Column(string name, string columnAlias, string tableAlias);
+        IQueryBuilder<TRecord> AllColumns();
+        IQueryBuilder<TRecord> CalculatedColumn(string expression, string columnAlias);
+        IQueryBuilder<TNewRecord> AsType<TNewRecord>() where TNewRecord : class;
 
+        IQueryBuilder<TRecord> AddRowNumberColumn(string columnAlias);
+        IQueryBuilder<TRecord> AddRowNumberColumn(string columnAlias, params string[] partitionByColumns);
+        IQueryBuilder<TRecord> AddRowNumberColumn(string columnAlias, params ColumnFromTable[] partitionByColumns);
+        IQueryBuilder<TRecord> Parameter(Parameter parameter);
+        IQueryBuilder<TRecord> ParameterDefault(Parameter parameter, object defaultValue);
+        IQueryBuilder<TRecord> Parameter(Parameter parameter, object value);
+
+        IJoinSourceQueryBuilder<TRecord> Join(IAliasedSelectSource source, JoinType joinType);
+        ISubquerySourceBuilder<TRecord> Union(IQueryBuilder<TRecord> queryBuilder);
+        ISubquerySourceBuilder<TRecord> Subquery(); 
+        ISelectBuilder GetSelectBuilder();
         int Count();
         bool Any();
         TRecord First();
         IEnumerable<TRecord> Take(int take);
         List<TRecord> ToList(int skip, int take);
         List<TRecord> ToList(int skip, int take, out int totalResults);
         List<TRecord> ToList();
-        void Delete();
         IEnumerable<TRecord> Stream();
         IDictionary<string, TRecord> ToDictionary(Func<TRecord, string> keySelector);
+        Parameters Parameters { get; }
```

Comment:
This and the property below are mutable. Is the intention that the consumer add their own parameters directly to these collections?

Reply:
No, consumers should not add directly to these collections. I exposed these because I need them to be public in order to create stored procs / views / functions in extension methods. I'm thinking that maybe these should actually be instance methods on the query builder to avoid `Parameters` being made public. Alternatively, I can expose `Parameters` immutably.

Reply:
You could make it `internal` if the extension methods are in the same assembly.

Reply:
True, I suppose that's the easiest option

Reply:
Actually this doesn't work. It is just an interface, so you can't make it internal.

Reply:
I'm thinking of leaving these as is. Moving stored procs / views / functions to instance methods doesn't mean that I can remove these, because there are other places that need them. For example, you still need to copy them to a new query builder by accessing them through the public `IQueryBuilder` interface when you add a join to your query.

The other option (changing the types exposed here to be immutable) doesn't feel like it adds a lot of value to me because we are already cloning the objects when you access these public properties. So as a consumer, you are unable to affect the query by mutating these objects.